runtime/src/iree-amd-aie/aie_runtime/iree_aie_runtime.cc

```diff
@@ -629,6 +629,135 @@ struct AMDAIEDeviceModel getDeviceModel(AMDAIEDevice device) {
+      {getElementTypeKey(8, 8, 32), {4, 8, 8}},
+
+      // `vector<4x4xi16>`  | `vector<4x8xi8>`   | `vector<4x8xi32>`
+      {getElementTypeKey(16, 8, 32), {4, 8, 4}},
+
+      // `vector<4x2xi16>`  | `vector<2x8xi16>`  | `vector<4x8xi32>`
+      {getElementTypeKey(16, 16, 32), {4, 8, 2}},
+
+      // `vector<2x8xi16>`  | `vector<8x8xi8>`   | `vector<2x8xi64>`
+      // `vector<4x8xi16>`  | `vector<8x4xi8>`   | `vector<4x4xi64>`
+      //   choosing the first i16 x i8 -> i64 instruction (arbitrarily)
+      {getElementTypeKey(16, 8, 64), {2, 8, 8}},
+
+      // `vector<2x4xi16>`  | `vector<4x8xi16>`  | `vector<2x8xi64>`
+      // `vector<4x4xi16>`  | `vector<4x4xi16>`  | `vector<4x4xi64>`
+      //   choosing the first i16 x i16 -> i64 instruction (arbitrarily)
+      {getElementTypeKey(16, 16, 64), {2, 8, 4}},
+
+      // `vector<4x2xi32>`  | `vector<2x4xi16>`  | `vector<4x4xi64>`
+      {getElementTypeKey(32, 16, 64), {4, 4, 2}},
+  };
+  return matmulIntSizes;
+}
+
+/// Return the AIE instruction size (m, n, k) for the integer types with
+/// bitwidths nBitsLhs, nBitsRhs, and nBitsAcc. Based on the table above.
+static llvm::FailureOr<std::array<uint32_t, 3>>
+getAIEIntegerMatmulInstructionSize(uint32_t nBitsLhs, uint32_t nBitsRhs,
+                                   uint32_t nBitsAcc) {
+  static llvm::DenseMap<uint32_t, std::array<uint32_t, 3>> &mapForIntTypes =
+      getIntegerMatmulInstructionSizeMap();
+  auto it =
+      mapForIntTypes.find(getElementTypeKey(nBitsLhs, nBitsRhs, nBitsAcc));
+  if (it == mapForIntTypes.end()) {
+    return failure();
+  }
+  return it->second;
+}
+
+llvm::FailureOr<std::array<uint32_t, 3>>
+AMDAIEDeviceModel::getAIEMatmulInstructionSize(Type elTypeLhs, Type elTypeRhs,
+                                               Type elTypeAcc) const {
+  bool allFloatingPoint = isa<FloatType>(elTypeLhs) &&
+                          isa<FloatType>(elTypeRhs) &&
+                          isa<FloatType>(elTypeAcc);
+
+  bool allInteger = isa<IntegerType>(elTypeLhs) &&
+                    isa<IntegerType>(elTypeRhs) && isa<IntegerType>(elTypeAcc);
+
+  if (!allInteger && !allFloatingPoint) {
+    return failure();
+  }
+
+  auto nBitsLhs = elTypeLhs.getIntOrFloatBitWidth();
+  auto nBitsRhs = elTypeRhs.getIntOrFloatBitWidth();
+  auto nBitsAcc = elTypeAcc.getIntOrFloatBitWidth();
+
+  if (allFloatingPoint) {
+    if (nBitsLhs == 16 && nBitsRhs == 16 && nBitsAcc == 32) {
+      if (device == AMDAIEDevice::npu4) {
+        // Strix nup4 intrinsics.
+        return std::array<uint32_t, 3>{8, 8, 8};
+      } else {
+        // Phoenix npu1_4col intrinsics.
+        return std::array<uint32_t, 3>{4, 4, 8};
+      }
+    }
+    // There is only 1 floating point case in the table (handled above).
+    return failure();
+  }
+
+  assert(allInteger &&
+         "expected all element types to either be all float types or all "
+         "integer types");
+  assert((device != AMDAIEDevice::npu4) &&
```

Comment:
Shouldn't this be: `device == AMDAIEDevice::npu1_4col ...` as integer types are only supported for npu1 for now.

Reply:
Yes, I have changed it now.